README.md
# Fork

* This library fork from [PracticalCrypto](https://github.com/gutierrezps/PracticalCrypto)
* The goal from this fork :
   - To encrypt/ decryipt using less memory by reducing String
   - Using existing buffer to encrypt/decrypt something
   - Encrypt/decrypt other than text

* Changes :
   - Encrypt using char array instead String
   - Decrypt using char array instead String
   - more info please read [CHANGELOG](CHANGELOG.md)
* GNU LGPL 2.1.

# PracticalCrypto

A simple library to encrypt and decrypt strings with hashing, for ESP8266, based
on [BearSSL](bearssl.org/) library that is already available on the ESP8266
Arduino core.

Based on "[Practical IoT Cryptography On The Espressif ESP8266](https://hackaday.com/2017/06/20/practical-iot-cryptography-on-the-espressif-esp8266/)",
by Sean Boyce on Hackaday, published at June 20, 2017.

Implementation details of this library are shown at <https://gutierrezps.wordpress.com/2020/05/25/practical-iot-cryptography-on-the-esp8266-using-arduino/>.

## Features

* Simple usage, only three main methods can be used
* String inputs and outputs
* Encryption key generation

## Examples

The `examples` folder contains library examples for Arduino. The `extras` folder
contains examples for PHP (run with `php <example>.php`) and NodeJS
(run with `nodejs <example>.js`).

## API

* `bool setKey(String key)` - Set an encryption key (exactly
64 char long), Returns true if key was set sucessfully set.
* `String getKey()` - Returns the key currently set.
* `String generateKey()` - Generates a random 64-char long
encryption key.
* `String encrypt(String plaintext)` - Encrypts the given
plaintext with the key set previously. Returns the ciphertext
encoded as an hex string.
* `String decrypt(String ciphertext)` - Decrypts the given
ciphertext hex string with the key set previously. Returns
the original plaintext.
* `uint16_t hexStringToArray(String input, uint8_t *output, uint16_t capacity)` -
    Converts an hex string to a byte array.
* `String arrayToHexString(uint8_t *input, uint16_t len)` - Converts a byte
    array to an hex string.
* `PracticalCrypto::Status lastStatus()` - If something went wrong while
    encrypting, decrypting or converting from hex to binary, `lastStatus()` will
    indicate what went wrong.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0.1] - 2020-10-22
### Added
* Encryption char array
* Decryption char array
* Add `arrayToHexCharArray` and `hexCharArrayToArray`
* Add `calculateBuffer` to calculate size of buffer to be use for any given size

### Changed
* Change flow of `encrypt` and `decrypt` from full String to char array
* Using `arrayToHexCharArray` to `encrypt` instead `arrayToHexString`
* Using `hexCharArrayToArray` to `decrypt` instead `hexStringToArray`


### Removed
* none

src/PracticalCrypto.cpp
/**
 * @file PracticalCrypto.cpp
 * @author Gutierrez PS (https://github.com/gutierrezps)
 *         Fahhrizal HU (https://github.com/k2biru)
 * @brief Library for easy encryption and decryption of Strings in ESP8266 Arduino core.
 * @version 0.1.1
 * @date 2020-10-20
 * 
 */
#include <Arduino.h>
#include <bearssl/bearssl_block.h>
#include <bearssl/bearssl_hmac.h>

#include "PracticalCrypto.h"

#define BLK_SZ   br_aes_big_BLOCK_SIZE


bool PracticalCrypto::setKey(String key)
{
    if (key.length() != 64) {
        key_ = "";
        return false;
    }

    key_ = key;

    key_.toCharArray((char*)staticIv_, 17, 0);
    key_.toCharArray((char*)ivKey_, 17, 16);
    key_.toCharArray((char*)dataKey_, 17, 32);
    key_.toCharArray((char*)hashKey_, 17, 48);

    return true;
}


String PracticalCrypto::generateKey()
{
    String key = "";
    for (uint8_t i = 0; i < 64; ++i) {
        key += (char)(32 + secureRandom(90));
    }
    return key;
}

/**
 * Calculate size of buffer to be use for any given plaint Size
 * 
 * buffer size is calculate from following parts, where 'n' is its plainSize:
 * - [0..31]  is the encrypted initialization vector
 * - [32..(n-41)] is the encrypted plaintext
 * - [(n-40)..(n-1)] is the SHA1 hash
 * 
 * @param plainSize     size of plainText
 * @return              cipher size
 */

size_t PracticalCrypto::calculateBuffer(const size_t plainSize){
    return (BLK_SZ+br_sha1_SIZE+(plainSize/ BLK_SZ + 1)*BLK_SZ)*2;
}

/**
 * Encrypts the provided plaintext String. The following steps are done:
 * 
 * 1. Plaintext is padded to be evenly divided into blocks of BLK_SZ bytes
 *      If it's already evenly divided, a whole padding block is added,
 *      The value of the byte used to pad is the number of padding bytes needed.
 * 2. A random initialization vector (IV) is generated
 * 3. The random IV is encryted (AES128-CBC) using ivKey_ and staticIv_
 * 4. The data is encrypted (AES128-CBC) using dataKey_ and the random IV
 * 5. The hash (HMAC-SHA1) of encrypted IV + encrypted data is calculated,
 *      using hashKey_ and following the Encrypt-then-MAC approach.
 * 6. The output String (ciphertext) is formed by concatenating hex strings of
 *      encrypted IV, encrypted data and hash.
 * 
 * @param plainText     String to be encrypted
 * @return String       ciphertext
 */
const String PracticalCrypto::encrypt(String &plainText)
{
    return encrypt(plainText.c_str(), plainText.length());
}

const String PracticalCrypto::encrypt(const char*plainText, const size_t plainSize){
    size_t cipherSize = calculateBuffer(plainSize);
    char *buffer =  reinterpret_cast<char*>(malloc(cipherSize+1));
    memcpy(buffer, plainText, plainSize);
    size_t size = encryptArray(buffer,plainSize,buffer);
    String ret;
    ret.reserve(size);
    if (size !=0){
        ret = buffer;
    }
    free(buffer);
    return ret;
}

const size_t PracticalCrypto::encryptArray(const char *plainText,size_t plainSize,char* cipher)
{
    size_t size = 0;
    if (key_.length() == 0) {
        lastStatus_ = InvalidKey;
        return size;
    }

    if (plainSize > kMaxDataLength_) {
        lastStatus_ = PlaintextTooLong;
        return size;
    }

    if (!dataBuffer_ || !cipher) {
        lastStatus_ = BufferAllocationFailed;
        return size;
    }
    // Serial.printf("here1 plainSize %i %s\n",plainSize,plainText);

    uint16_t i = 0;

    // dataBuffer_ will turn into the encrypted plaintext later
    memcpy(dataBuffer_, plainText, plainSize);

    
    // number of blocks required for data is rounded up if division has decimals,
    // otherwise a full padding block is added if the division is even
    uint8_t dataBlocksQty = plainSize / BLK_SZ + 1;

    // add padding to dataBuffer_
    uint8_t dataPadding = dataBlocksQty * BLK_SZ - plainSize;
    const uint16_t paddingStart = dataBlocksQty * BLK_SZ - dataPadding;
    for (i = paddingStart; i < dataBlocksQty * BLK_SZ; ++i) {
        dataBuffer_[i] = dataPadding;
    }

    // generating a random IV of printable chars (ASCII 32-122)
    uint8_t iv[BLK_SZ] = {0};
    for (i = 0; i < BLK_SZ-1; ++i) {
        iv[i] = 32 + secureRandom(90);
    }

    // manual padding is added in order to encrypted IV be exactly BLK_SZ bytes;
    // otherwise, when decrypting in PHP a whole padding block would be expected
    iv[BLK_SZ-1] = 1;

    yield();    // reset watchdog

    // encryption context
    br_aes_big_cbcenc_keys encCtx;

    // encrypt IV
    uint8_t staticIv[BLK_SZ] = {0};
    memcpy(staticIv, staticIv_, BLK_SZ);      // copy to local, modifiable variable

    uint8_t ivCipher[BLK_SZ] = {0};
    memcpy(ivCipher, iv, BLK_SZ);

    br_aes_big_cbcenc_init(&encCtx, ivKey_, BLK_SZ);
    br_aes_big_cbcenc_run(&encCtx, staticIv, ivCipher, BLK_SZ);

    yield();    // reset watchdog

    // reset the encryption context and encrypt the data
    br_aes_big_cbcenc_init(&encCtx, dataKey_, BLK_SZ);
    br_aes_big_cbcenc_run(&encCtx, iv, dataBuffer_, dataBlocksQty * BLK_SZ);

    yield();

    // calculate hash

    // contexts
    br_hmac_key_context hashKc;
    br_hmac_context hashCtx;


    // initialize key context with the SHA1 algorithm, the given key and its length
    br_hmac_key_init(&hashKc, &br_sha1_vtable, hashKey_, BLK_SZ);


    // initialize hashing context, setting the output size
    br_hmac_init(&hashCtx, &hashKc, br_sha1_SIZE);


    // hash encrypted IV and encrypted data
    br_hmac_update(&hashCtx, ivCipher, BLK_SZ);
    br_hmac_update(&hashCtx, dataBuffer_, dataBlocksQty * BLK_SZ);


    yield();

    // get the output hash
    uint8_t hash[br_sha1_SIZE] = {0};
    br_hmac_out(&hashCtx, hash);


    size = (arrayToHexCharArray(ivCipher,BLK_SZ,cipher,size));
    size = (arrayToHexCharArray(dataBuffer_,dataBlocksQty * BLK_SZ,cipher,size));
    size = (arrayToHexCharArray(hash, br_sha1_SIZE,cipher,size));

    lastStatus_ = Ok;

    return size;
}


/**
 * Decrypts the provided ciphertext String. The following steps are done:
 * 
 * 1. Ciphertext is validated and converted to byte arrays.
 * 2. Hash of encrypted IV + encrypted data is calculated using hashKey_
 *      and compared with hash extracted from ciphertext
 * 3. IV is decrypted using ivKey_ and staticIv_
 * 4. Data is decrypted using IV and dataKey_
 * 5. Output string (plaintext) is generated from data, excluding padding bytes
 * 
 * @param ciphertext    String to be decrypted
 * @return String       plaintext data
 */
const String PracticalCrypto::decrypt(String &cipherText)
{
    return decrypt(cipherText.c_str(),cipherText.length());
}

/**
 * Decrypts the provided ciphertext array. The following steps are done:
 * 
 * 1. Ciphertext is validated and converted to byte arrays.
 * 2. Hash of encrypted IV + encrypted data is calculated using hashKey_
 *      and compared with hash extracted from ciphertext
 * 3. IV is decrypted using ivKey_ and staticIv_
 * 4. Data is decrypted using IV and dataKey_
 * 5. Output string (plaintext) is generated from data, excluding padding bytes
 * 
 * @param ciphertext    encrypted const char*
 * @param size          size of ciphertext
 * @return String       plaintext, empty string if failed
 */
const String PracticalCrypto::decrypt(const char* cipherText, size_t size)
{   
    String out;
    out.reserve(size);
    char buffer[size];
    memcpy(buffer,cipherText,size);
    if(!decryptArray(buffer,size)){
        return out;
    }
    out = String(buffer);
    return out;
}

/**
 * Decrypts the provided data array ciphertext. The following steps are done:
 * 
 * 1. Ciphertext is validated and converted to byte arrays.
 * 2. Hash of encrypted IV + encrypted data is calculated using hashKey_
 *      and compared with hash extracted from ciphertext
 * 3. IV is decrypted using ivKey_ and staticIv_
 * 4. Data is decrypted using IV and dataKey_
 * 5. Output plaintext save to data if decrypts successfully, excluding padding bytes
 * 
 * @param data          buffer (array) of data ciphertext and if decrypt plaintext successfully 
 * @param cipherSize    size of cipher
 * @return size_t       return size of plaintext in data array, return 0 if failed
 */
const size_t PracticalCrypto::decryptArray(char* data, size_t cipherSize)
{   
    size_t decSize = 0;
    if (key_.length() == 0) {
        lastStatus_ = InvalidKey;
        return decSize;
    }

    if (!dataBuffer_) {
        lastStatus_ = BufferAllocationFailed;
        return decSize;
    }

    // minimum ciphertext hex string length:
    // iv + min plaintext length (padded) + sha1 output length
    const uint16_t minCiphertextLength = (16 + 16 + 20)*2;

    if (cipherSize < minCiphertextLength || cipherSize % 2 != 0) {
        lastStatus_ = InvalidCiphertextLength;
        return decSize;
    }

    // maximum ciphertext hex string length:
    // iv + max plaintext length (padded) + sha1 output length
    const uint16_t maxCiphertextLength = (16 + (kMaxDataLength_ + 16) + 20)*2;

    if (cipherSize > maxCiphertextLength) {
        lastStatus_ = CiphertextTooLong;
        return decSize;
    }

    const uint16_t ivEnd = BLK_SZ*2;
    const uint16_t hashStart = cipherSize - br_sha1_SIZE*2;


    uint16_t converted = 0;
    const uint16_t dataLength = (hashStart-ivEnd) / 2;

    uint8_t ivCipher[BLK_SZ] = {0};
    uint8_t hashCipher[br_sha1_SIZE] = {0};

    converted = hexCharArrayToArray((char*)data,0,ivEnd, ivCipher, BLK_SZ);
    if (converted == 0) {
        // last status already set
        return decSize;
    }

    converted = hexCharArrayToArray((char*)data,ivEnd, hashStart, dataBuffer_, dataLength);
    if (converted == 0) {
        // last status already set
        return decSize;
    }

    converted = hexCharArrayToArray((char*)data,hashStart,cipherSize,hashCipher,br_sha1_SIZE);
    if (converted == 0) {
        // last status already set
        return decSize;
    }

    yield();

    // calculate hash

    // contexts
    br_hmac_key_context hashKc;
    br_hmac_context hashCtx;

    // initialize key context with the SHA1 algorithm, the given key and its length
    br_hmac_key_init(&hashKc, &br_sha1_vtable, hashKey_, BLK_SZ);

    // initialize hashing context, setting the output size
    br_hmac_init(&hashCtx, &hashKc, br_sha1_SIZE);

    // hash encrypted IV and encrypted data
    br_hmac_update(&hashCtx, ivCipher, BLK_SZ);
    br_hmac_update(&hashCtx, dataBuffer_, dataLength);

    yield();

    // get the hash value
    uint8_t hashExpected[br_sha1_SIZE] = {0};
    br_hmac_out(&hashCtx, hashExpected);

    if (memcmp(hashExpected, hashCipher, br_sha1_SIZE) != 0) {
        lastStatus_ = HashMismatch;
        return decSize;
    }

    // decryption context
    br_aes_big_cbcdec_keys decCtx;

    uint8_t staticIv[BLK_SZ] = {0};
    memcpy(staticIv, staticIv_, BLK_SZ);

    // decrypt IV
    br_aes_big_cbcdec_init(&decCtx, ivKey_, BLK_SZ);
    br_aes_big_cbcdec_run(&decCtx, staticIv, ivCipher, BLK_SZ);

    yield();

    // decrypt data
    br_aes_big_cbcdec_init(&decCtx, dataKey_, BLK_SZ);
    br_aes_big_cbcdec_run(&decCtx, ivCipher, dataBuffer_, dataLength);

    yield();
    
    // get number of padding bytes used
    const uint8_t dataPadding = dataBuffer_[dataLength - 1];
    
    decSize = dataLength - dataPadding;

    // insert a null char to terminate the string
    dataBuffer_[decSize] = 0;

    lastStatus_ = Ok;

    // copy paintext to data
    memcpy(data, dataBuffer_,decSize);
    // add null char to terminate the string
    data[decSize] = 0;
    
    return decSize;
}




inline int8_t hexToByte(char hex)
{
    if (hex >= '0' && hex <= '9') {
        return hex - '0';
    }
    else if (hex >= 'a' && hex <= 'f') {
        return hex - 'a' + 10;
    }
    else if (hex >= 'A' && hex <= 'F') {
        return hex - 'A' + 10;
    }
    
    return -1;
}


uint16_t PracticalCrypto::hexCharArrayToArray(
    char * input,
    uint16_t inputStart,
    const uint16_t inputStop,
    uint8_t *output,
    const uint16_t capacity)
{
    uint8_t val = 0;
    uint16_t i = 0;
    char ch = 0;
    uint16_t bytesQty = (inputStop-inputStart) / 2;

    if (((inputStop-inputStart) % 2) != 0) {
        lastStatus_ = InvalidHexString;
        return 0;
    }

    if (bytesQty > capacity) {
        lastStatus_ = HexStringTooLong;
        return 0;
    }
    if(inputStart!=0){
        inputStart = inputStart/2;
    }
    for (i = 0; i < bytesQty; ++i) {
        val = 0;
        ch = input[inputStart*2];
        if (hexToByte(ch) < 0) {
            lastStatus_ = InvalidHexString;
            return 0;
        }
        val += hexToByte(ch);
        val *= 16;
        ch = input[(inputStart * 2) + 1];
        if (hexToByte(ch) < 0) {
            lastStatus_ = InvalidHexString;
            return 0;
        }
        val += hexToByte(ch);
        output[i] = val;
        inputStart ++;
    }
    lastStatus_ = Ok;
    return i;
}


const String PracticalCrypto::arrayToHexString(uint8_t *input, uint16_t len)
{
    String ret = "";

    for (uint16_t i = 0; i < len; ++i) {
        char ch = (input[i] >> 4) & 0x0F;
        if (ch < 10) ch += '0';
        else ch += 'A' - 10;
        ret += ch;

        ch = input[i] & 0x0F;
        if (ch < 10) ch += '0';
        else ch += 'A' - 10;
        ret += ch;
    }
    

    return ret;
}

const size_t PracticalCrypto::arrayToHexCharArray(uint8_t *input, size_t inputLen, char *output, size_t outputStart)
{
    size_t size = 0;
    size_t sizeOffset = outputStart;
    for (uint16_t i = 0; i < inputLen; ++i) {
        size = (i*2)+sizeOffset;
        char ch = (input[i] >> 4) & 0x0F;
        if (ch < 10) ch += '0';
        else ch += 'A' - 10;
        output[size] = ch;

        size++;

        ch = input[i] & 0x0F;
        if (ch < 10) ch += '0';
        else ch += 'A' - 10;
        output[size] = ch;
    }
    size++;
    output[size] = 0;
    return size;
}

#undef BLK_SZ


src/PracticalCrypto.h
/**
 * @file PracticalCrypto.cpp
 * @author Gutierrez PS (https://github.com/gutierrezps)
 *         Fahhrizal HU (https://github.com/k2biru)
 * @brief Library for easy encryption and decryption of Strings in ESP8266 Arduino core.
 * @version 0.1.1
 * @date 2020-10-20
 * 
 * Library for easy encryption and decryption of Strings in ESP8266 Arduino core.
 * 
 * Based on "Practical IoT Cryptography On The Espressif ESP8266",
 * by Sean Boyce on Hackaday, published at June 20, 2017.
 * https://hackaday.com/2017/06/20/practical-iot-cryptography-on-the-espressif-esp8266/
 * 
 * Uses BearSSL library, included in ESP8266 Arduino Core:
 * - Block Cypher (AES128 used): https://bearssl.org/apidoc/bearssl__block_8h.html
 * - HMAC (SHA1 used): https://bearssl.org/apidoc/bearssl__hmac_8h.html
 * 
 */

#ifndef PracticalCrypto_h
#define PracticalCrypto_h

#include "Arduino.h"

/**
 * @brief String encryption and decryption class with hashing included.
 * 
 * Uses AES128-CBC for encryption and HMAC-SHA1 for hashing.
 * After setting a valid encryption key with setKey(), simply call encrypt()
 * or decrypt(). No additional configuration is necessary.
 * 
 */
class PracticalCrypto {
public:
    /**
     * @brief Possible status for encryption, decryption and hexStringToArray
     */
    enum Status : char {
        Ok = 0,
        InvalidKey,
        InvalidCiphertextLength,
        InvalidHexString,
        PlaintextTooLong,
        CiphertextTooLong,
        HexStringTooLong,
        HashMismatch,
        BufferAllocationFailed
    };
    
    PracticalCrypto(const uint16_t maxDataLength = 1024) : kMaxDataLength_(maxDataLength) {
        dataBuffer_ = (uint8_t*) malloc(kMaxDataLength_ + 16);
    }

    /**
     * @brief Set the encryption key. Must constain exactly 64 chars, where:
     * - [0..15]  is the static initialization vector
     * - [16..31] is the initialization vector encryption key
     * - [32..47] is the data encryption key
     * - [48..63] is the hashing key
     * 
     * If an invalid key is provided, the current key will be reset, so
     * encryption and decryption will not be possible until a valid
     * key is set.
     * 
     * @param key   String with exactly 64 chars
     * @return      true if key has the right length
     */
    bool setKey(String key);

    /**
     * @return Current encryption key
     */
    String getKey() const {
        return key_;
    }

    /**
     * @brief Generates a random key of 64 printable chars (ASCII 32-122)
     * 
     * @return key string
     */
    String generateKey();

    /**
     * @brief Encrypt the provided plaintext String using the key set, 
     * and returns the ciphertext encoded as an hex string.
     * 
     * The key must be sucessfully set before trying to encrypt.
     * 
     * An empty string is returned in case of error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, PlaintextTooLong
     * or Ok if encrypted correctly.
     * 
     * Ciphertext is composed of the following parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * @param plainText     String to be encrypted, must be shorter than maxLength
     * @return              ciphertext, empty string if failed
     */
    const String encrypt(String &plainText);

    /**
     * @brief Encrypt the provided plaintext String using the key set, 
     * and returns the ciphertext encoded as an hex string.
     * 
     * The key must be sucessfully set before trying to encrypt.
     * 
     * An empty string is returned in case of error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, PlaintextTooLong
     * or Ok if encrypted correctly.
     * 
     * Ciphertext is composed of the following parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * @param plainText     Array to be encrypted, must be shorter than maxLength
     * @param plainSize     size of plainText
     * @return              ciphertext, empty string if failed
     */
    const String encrypt(const char*plainText, const size_t plainSize);

    /**
     * @brief Encrypt the provided plainText Array using the key set, 
     * and returns cipher Size encoded as an hex string.
     * 
     * The key must be sucessfully set before trying to encrypt.
     * 
     * return 0 if there any error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, PlaintextTooLong
     * or Ok if encrypted correctly.
     * 
     * Ciphertext is composed of the following parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * @param plainText     Array to be encrypted, must be shorter 
     *                      than maxLength
     * @param plainSize     size of plainText
     * @param cipher        char * to be save as cipher
     * @return              cipher size, 0 if failed
     */
    const size_t encryptArray(const char *plainText,size_t plainSize,char* cipher);

    /**
     * @brief Calculate size of buffer to be use for any given plaint Size
     * 
     * buffer size is calculate from following parts, where 'n' is its plainSize:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * @param plainSize     size of plainText
     * @return              cipher size
     */
    size_t calculateBuffer(const size_t plainSize);

    /**
     * Decrypt the provided ciphertext String using the key set,
     * and returns the plaintext.
     * 
     * The key must be sucessfully set before trying to decrypt.
     * 
     * An empty string is returned in case of error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, InvalidCiphertextLength,
     * InvalidHexString, HashMismatch, CiphertextTooLong, or Ok
     * if decrypted correctly.
     * 
     * Ciphertext is interpreted as an hex string composed of the following
     * parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * Therefore, ciphertext must have an even number of chars, be at least
     * 104 chars long, and encrypted plaintext length must be a multiple of 32,
     * since the encryption block size is 16.
     * 
     * @param ciphertext    encrypted String
     * @return String       plaintext, empty string if failed
     */
    const String decrypt(String &cipherText);

     /**
     * @brief Decrypt the provided ciphertext array using the key set,
     * and returns the plaintext String.
     * 
     * The key must be sucessfully set before trying to decrypt.
     * 
     * An empty string is returned in case of error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, InvalidCiphertextLength,
     * InvalidHexString, HashMismatch, CiphertextTooLong, or Ok
     * if decrypted correctly.
     * 
     * Ciphertext is interpreted as an hex string composed of the following
     * parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * Therefore, ciphertext must have an even number of chars, be at least
     * 104 chars long, and encrypted plaintext length must be a multiple of 32,
     * since the encryption block size is 16.
     * 
     * @param ciphertext    encrypted const char*
     * @param size          size of ciphertext
     * @return String       plaintext, empty string if failed
     */
    const String decrypt(const char* cipherText, size_t size);
    
    /**
     * @brief Decrypt the provided data array ciphertext using the key set,
     * and returns size of plaintext.
     * 
     * The key must be sucessfully set before trying to decrypt.
     * 
     * An empty size is returned in case of error. Possible statuses
     * are BufferAllocationFailed, InvalidKey, InvalidCiphertextLength,
     * InvalidHexString, HashMismatch, CiphertextTooLong, or Ok
     * if decrypted correctly.
     * 
     * Ciphertext is interpreted as an hex string composed of the following
     * parts, where 'n' is its length:
     * - [0..31]  is the encrypted initialization vector
     * - [32..(n-41)] is the encrypted plaintext
     * - [(n-40)..(n-1)] is the SHA1 hash
     * 
     * Therefore, data array must have an even number of chars, be at least
     * 104 chars long, and encrypted plaintext length must be a multiple of 32,
     * since the encryption block size is 16.
     * 
     * @param data          buffer (array) of data ciphertext and if decrypt plaintext successfully
     * @param cipherSize    size of cipher
     * @return size_t       return size of plaintext in data array, return 0 if failed
     */
    const size_t decryptArray(char* data, size_t cipherSize);

    /**
     * @brief Get last status, set by encrypt or decrypt methods.
     * 
     * @return Status
     */
    Status lastStatus() const {
        return lastStatus_;
    }

    /**
     * @brief Converts an hex char array to a byte array. In case of error,
     * 0 is returned and lastStatus() will indicate what happened.
     * 
     * Possible errors are InvalidHexString (invalid chars detected)
     * or HexStringTooLong (input larger than capacity).
     * 
     * @param  input        length must be even and must fit into capacity
     * @param  inputStart   input start index
     * @param  inputStop    input stop index
     * @param  output       output byte array
     * @param  capacity     output capacity
     * @return              number of bytes converted, 0 in case of error
     */
    uint16_t hexCharArrayToArray(char *input, uint16_t inputStart, const uint16_t inputStop, uint8_t *output, const uint16_t capacity);

    /**
     * @brief Converts a byte array to an hex string.
     * 
     * @param  input    byte array to be converted
     * @param  len      byte array length
     * @return      hex string
     */
    const String arrayToHexString(uint8_t *input, uint16_t len);

    /**
     * @brief Converts a byte array to an hex char array.
     * 
     * @param  input        byte array to be converted
     * @param  inputLen     byte array length
     * @param  output       pointer to output array
     * @param  outputStart  output array start index
     * @return              size of output
     */
    const size_t arrayToHexCharArray(uint8_t *input, size_t inputLen, char *output, size_t outputStart);


private:
    String key_ = "";
    Status lastStatus_ = Status::Ok;
    uint8_t *dataBuffer_ = nullptr;
    const uint16_t kMaxDataLength_;
    uint8_t staticIv_[17] = {0};
    uint8_t ivKey_[17] = {0};
    uint8_t dataKey_[17] = {0};
    uint8_t hashKey_[17] = {0};
};


#endif